### PR TITLE
Optimize memory utilization by keeping logits in BF16.

### PR DESCRIPTION
**Pull request by kalyank007**
Use the `--logits_bf16` parameter to keep logits in BF16; otherwise, fall back to the default path.

### CODE REVIEW

**Comment by yafshar** [comment]
@kalyank007 `logits.float()` is removed in most ModelForCausalLM implementations to save memory. Do you know why in the case of mllama, the logits still kept in float? Is there a specific requirements or optimizations related to its architecture. Does your change cause an issue?

**Comment by regisss** [comment]
@yafshar We are keeping the logits in fp32 for a few models in the Transformers 4.49 PR because we noticed some minor text-generation regressions. So, I can't speak for @kalyank007, but I can see why an option to choose the dtype of the logits is useful.

**Comment by kalyank007** [comment]
> @kalyank007 `logits.float()` is removed in most ModelForCausalLM implementations to save memory. Do you know why in the case of mllama, the logits still kept in float? Is there a specific requirements or optimizations related to its architecture. Does your change cause an issue?

In the case of mllama, logits are still kept in float likely due to specific architectural requirements or optimizations, such as maintaining numerical stability during post-processing steps like sampling or loss computation.

Regarding the introduced change, it is protected by the **logits_bf16** flag, ensuring that the current code flow remains unchanged unless explicitly enabled. This modification allows us to save memory and scale to higher batch sizes, particularly benefiting the LLaMA 3.2 90B model. The change is tested for LLaMA 3.2 90B model and produces the same text output with no accuracy degradation.

**Review by yafshar — APPROVED** [review]
LGTM!

@regisss this PR is ready. would you please do your final check on this PR

**Review by regisss — APPROVED** [review]
LGTM!

The failed CI test is unrelated.

**Comment by regisss** [comment]
I will just wait for #1698 to be merged before merging this one

**Comment by regisss** [comment]
@kalyank007 Can you solve the merge conflicts following #1698 please?

**Comment by kalyank007** [comment]
> @kalyank007 Can you solve the merge conflicts following #1698 please?

Done!

**Review by regisss — APPROVED** [review]
